src/backend/columnar/columnar_reader.c

```diff
@@ -444,6 +444,17 @@ ColumnarReadChunkGroupsFiltered(ColumnarReadState *state)
 }
 
 
+/*
+ * ColumnarReadProjectedColumnList returns a list of Var objects for the
+ * columns that are projected during this read operation.
+ */
+const List *
+ColumnarReadProjectedColumnList(ColumnarReadState *state)
+{
+	return state->projectedColumnList;
+}
```

Comment:
It seems like we are doing too much work here. The needed attributes list is generated in columnar_customscan.c in the first place, so why do we pass it all the way down to columnar_reader.c and then all the way back?

We can simply generate the needed columns the same way using ColumnarAttrNeeded. That returns a Bitmapset, but that's fine because all you need is the attr number anyway, you don't need the Var. This also has the benefit that it works for ordinary EXPLAIN without requiring ANALYZE (to generate the scan descriptor).